### PR TITLE
unusued variables

### LINKED ISSUE #37: Catch unused variables
At least "variable X is never written to" and "variable X is never read"

### PR DESCRIPTION
**Pull request by FredTheDino**
(Added in a bunch of extra stuff, just because it made the PR better... 3432a32 and forrward is relevant.)

Makes it a compilation error if a variable is unused.

I'm unsure how I feel about this currently. It can get annoying,
but I think it's a good help and it needs to be available somehow.
The way to silence the error is to just, read the value on the line
after. It might be better to enforce a naming standard, like Rust?
Or maybe this is a good way to do it, since you don't have to rename
stuff?

Closes #37

Starts work on #63

### CODE REVIEW

**Comment by FredTheDino** [comment]
killer feature: Allow the typechecker to run on code that is only faulty when it comes to unused variables. So a program can now be valid in different ways. Off to bed I go!

File: src/compiler.rs [hunk 1]
```diff
@@ -241,10 +265,27 @@ macro_rules! push_frame {
...
+                        var.line,
+                        Token::Identifier(var.name.clone()
+                    ));
+                    $compiler.error_on_line(
+                        e,
+                        var.line,
+                        Some(format!("Usage of undefined value: '{}'.", var.name))
+                    );
+                }
+                $compiler.panic = false;
```
Comment (sornas) [inline src/compiler.rs:287]: Maybe we'd like a proper warning!-macro instead of resetting the panic. Then it'd be easier to output "5 warnings found" or something.

File: src/compiler.rs [hunk 2]
```diff
@@ -329,6 +394,26 @@ impl Compiler {
         &mut self.frames[last]
     }
 
+    /// Marks a variable as read, also marks upvalues.
```
Comment (sornas) [inline src/compiler.rs:397]:
```suggestion
    /// Marks a variable as read. Also marks upvalues.
```

**Review by sornas — COMMENTED** [review]
n> It might be better to enforce a naming standard, like Rust?
> Or maybe this is a good way to do it, since you don't have to rename
> stuff?

Do you mean something like the variable "_x" not being warned about? I'm unsure, actually. I wonder if we'd need to apply code inspection in more places than just here for it to feel like it fits. But it's a neat idea.

As for the renaming part I think it's not much of an issue. If it's unused you only need to change it in one place and it's not like there's a lot of names to choose from.

File: src/compiler.rs (hunk 1 above)
Comment (sornas) [reply src/compiler.rs:287]: Out of scope though.

**Comment by FredTheDino** [comment]
> I think _some_ way of silencing the warning is needed before merging. Either renaming to `_x` or passing a hard-coded flag (as in, we don't need a full-blown `-Wno-unusued-variable`-system).

Not sure what the best solution is here. Passing flags to the compiler, will require a pretty large rewrite, and might be the better way. In some ways I dislike the Rust way, since it promotes a lot of edits to the code. But then again I guess people like it, because it makes it painfully obvious if a variable is unused.

Go enforces this as an error, making programs with unused variables faulty programs. We're not more extreme than Go here. And I think it's a good mentality to be extreme, and push for the features we want as hard as possible.

TL;DR I don't agree that we need a flag or way to silence the warning, other than making a "fake-usage" of the variable. We need to take a discussion here.

**Review by sornas — APPROVED** [review]
